components/layout_2020/table/construct.rs

```diff
@@ -0,0 +1,438 @@
+                    .expect(
+                        "Nonzero incoming rowspan cannot occur without a cell spanning this slot",
+                    )
+            } else {
+                TableSlot::Empty
+            };
+
+            self.table.push_new_slot_to_last_row(new_cell);
+            current_x = self.current_x();
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#algorithm-for-processing-rows
+    /// Push a single cell onto the slot map, handling any colspans it may have, and
+    /// setting up the outgoing rowspans.
+    pub fn add_cell(&mut self, cell: TableSlotCell) {
+        // Make sure the incoming_rowspans table is large enough
+        // because we will be writing to it.
+        let current_x = self.current_x();
+        let colspan = cell.colspan;
+        let rowspan = cell.rowspan;
+
+        if self.incoming_rowspans.len() < current_x + colspan {
+            self.incoming_rowspans.resize(current_x + colspan, 0isize);
+        }
+
+        debug_assert_eq!(
+            self.incoming_rowspans[current_x], 0,
+            "consume_rowspans must have been called before this!"
+        );
+
+        // If `rowspan` is zero, this is automatically negative and will stay negative.
+        let outgoing_rowspan = rowspan as isize - 1;
+        self.table.push_new_slot_to_last_row(TableSlot::Cell(cell));
+        self.incoming_rowspans[current_x] = outgoing_rowspan;
+
+        // Draw colspanned cells
+        for colspan_offset in 1..colspan {
+            let current_x_plus_colspan_offset = current_x + colspan_offset;
+            let new_offset = TableSlotOffset::new(colspan_offset, 0);
+            let incoming_rowspan = &mut self.incoming_rowspans[current_x_plus_colspan_offset];
+            let new_slot = if *incoming_rowspan == 0 {
+                *incoming_rowspan = outgoing_rowspan;
+                TableSlot::new_spanned(new_offset)
+            } else {
+                // This means we have a table model error.
+
+                // if `incoming_rowspan` is greater than zero, a cell from above is spanning
+                // into our row, colliding with the cells we are creating via colspan. In
+                // that case, set the incoming rowspan to the highest of two possible
+                // outgoing rowspan values (the incoming rowspan minus one, OR this cell's
+                // outgoing rowspan).  `spanned_slot()`` will handle filtering out
+                // inapplicable spans when it needs to.
+                //
+                // If the `incoming_rowspan` is negative we are in `rowspan=0` mode, (i.e.
+                // rowspan=infinity), so we don't have to worry about the current cell
+                // making it larger. In that case, don't change the rowspan.
+                if *incoming_rowspan > 0 {
+                    *incoming_rowspan = std::cmp::max(*incoming_rowspan - 1, outgoing_rowspan);
```

Comment:
We need to update `incoming_rowspans` here because `create_slots_for_cells_above_with_rowspan` hasn't yet been called for this column (we had stopped at the first cell opportunity) and it will never be called because adding a cell here increments `current_x` past this position. I've added a test to verify this behavior.